nexus/go-handler.md

```diff
@@ -0,0 +1,307 @@
+# Temporal Go SDK - Nexus Handler
+
+## Overview
+
+An addition to the [Temporal Go SDK][temporal-go-sdk] that exposes the ability for users to handle Nexus API requests.
+
+**Goals**
+
+- Expose a simple, high level API for binding Nexus operations to common Temporal constructs
+- Support for starting and canceling operations
+- Support multi-tenant applications with minimal boilerplate
+
+**Non-Goals**
+
+- Support for getting an operation's result or info
+- API definitions for other SDKs
+- API for invoking Nexus calls - refer to the [Nexus Go SDK][nexus-go-sdk] for an out-of-workflow reference client
+- Server implementation details
+- Worker / server API contract
+- Allow advanced users to implement Nexus inbound and outbound request proxies for cross-organization calls where
+  data converters are not shared and data is not exposed to Temporal server.
+
+  This will quickly become a requirement for production use cases, especially with Temporal Cloud's zero trust model. It
+  will be addressed shortly after Nexus becomes available for preview.
+
+> NOTE: These are non-goals for _this_ proposal, but they will be defined in follow up proposals / PRs.
+
+## Background
+
+Nexus is a thin protocol on top of HTTP, defined in [this spec][nexus-http-api]. Temporal serves as an HTTP proxy,
+translating Nexus requests into tasks that can be polled on by an SDK worker on its configured task queue.
+Due to the synchronous nature of HTTP, the generated tasks are not persisted, and require an available worker
+to be processed.
+
+The process of translating requests into tasks and delivering them to workers is out of scope for this proposal.
+
+## SDK Additions
+
+The Temporal Go SDK exposes a `Worker` interface that embeds a `Registry` interface defined as:
+
+```go
+type Registry interface {
+	WorkflowRegistry
+	ActivityRegistry
+}
+```
+
+Add and embed a third registry type: `OperationRegistry` with the following method:
+
+```go
+type OperationRegistry interface {
+	RegisterOperation(operation string, handler operation.Handler)
+	// Alternative: get the operation name from handler.Name()
+	// RegisterOperation(handler operation.Handler)
+}
+```
+
+`operation.Handler` is an interface with a minimal definition of:
+
+```go
+package operation // go.temporal.io/sdk/operation
+
+type Handler interface {
+	StartOperation(context.Context, *StartOperationRequest) (StartOperationResponse, error)
+	CancelOperation(context.Context, *CancelOperationRequest) error
+	// Later we will add GetResult and GetInfo as well
+}
+
+// Maps almost 1:1 to the Nexus Go SDK StartOperationRequest, exposing HTTP header and body directly instead of the HTTP
+// Request object.
+type StartOperationRequest struct {
+	Operation   string
+	RequestID   string
+	CallbackURL string
+
+	URL    *url.URL
+	Header http.Header
+	Body   []byte
```

Comment:
Hrmm, I wonder if a user might want HTTP method or TLS details. Can wait, but alternatively could reconstruct an `*http.Request` via gRPC response and add things on the gRPC response to build it up as needed.

In fact, I think just reusing the `nexus.Handler` interface is cleaner, even if you are proxying through the server.

Reply:
HTTP method is always POST for start, that's what we defined in the Nexus HTTP spec.
Not sure what you mean by gRPC response, Nexus doesn't use gRPC, gRPC is just for delivering Nexus tasks and completing tasks between frontend and worker (and internally matching).

But I may reuse the nexus.Handler interface and construct an http.Request as you mentioned, although I'm not 100% there's too much value there.

Reply:
> HTTP method is always POST for start, that's what we defined in the Nexus HTTP spec.

True, but what about TLS details? May not be needed now, but we should probably not just build our own form of HTTP request for every language if the language has a perfectly good one

> Not sure what you mean by gRPC response

I mean it's not a real HTTP request coming to the handler, it's built based on things that come from Temporal server over gRPC

> But I may reuse the nexus.Handler interface and construct an http.Request as you mentioned, although I'm not 100% there's too much value there.

I think there's plenty of value in developing a Nexus handler that can work with and without Temporal. So you should also probably provide a way to set the client in the context too.

Reply:
> I mean it's not a real HTTP request coming to the handler, it's built based on things that come from Temporal server over gRPC

Is there anything specific that you would add to the request from the PollNexusTaskResponse or the gRPC metadata? I was going to reflect the incoming HTTP request here, not the internal server-worker protocol.

I'll switch to use the nexus handler interface.

Reply:
Nah, I think we can start basic with the URL, header, and body like you have now and grow as needed

Reply:
Cool, I'll change then.